File: codigo/descriptive/02_calculate_road_density_departments.R


# 1. Set up environment -----
rm(list = ls())
gc()
if(!require(pacman)){install.packages('pacman');library(pacman)}
p_load(
  tidyverse,
  units,
  sf,
  geoarrow,
  purrr,
  furrr,
  future,
  parallel
)

# 2. Import data -----
pop_hex <- sf::read_sf(
  "datos/population/colombia/kontur_population_CO_20231101.gpkg")
col_dpto <- sf::read_sf(
  "datos/spatial/MGN2023_DPTO_POLITICO/MGN_ADM_DPTO_POLITICO.shp")
roads_colombia <- sf::read_sf("datos/spatial/colombia_roads.gpkg")


#=======================#
# 3. Road density dpto ------
#=======================#

pop_hex <- st_transform(pop_hex, crs = 4686)
roads_colombia <- st_transform(roads_colombia, st_crs(col_dpto))

if(!file.exists("datos/spatial/roads_dpto_intersect.rds")){
  # Prepare parallelization
  plan(multisession, workers = parallel::detectCores()-1)  
  
  intersections <- st_intersects(x = roads_colombia, y = col_dpto)
  
  # Initialize progress bar
  pb <- progress::progress_bar$new(format = "[:bar] :current/:total (:percent)", total = dim(roads_colombia)[1])
  
  # Define parallelized intersection function
  intersectFeatures <- future_map_dfr(1:dim(roads_colombia)[1], function(ix){
    pb$tick()
    st_intersection(x = roads_colombia[ix,], y = col_dpto[intersections[[ix]],])
  })
  plan(sequential)  # Reset the plan to the default single-core mode
  saveRDS(intersectFeatures, 
          file ="datos/spatial/roads_dpto_intersect.rds" ) 
  roads_by_department <- intersectFeatures
  
} else{
  
  roads_by_department <- readRDS("datos/spatial/roads_dpto_intersect.rds")
}

# Calculate the length of roads within each department #
roads_by_department$length_km <- st_length(roads_by_department) %>% 
  units::set_units("km") # Convert to kilometers

# Calculate department areas in square kilometers
col_dpto$area_km2 <- st_area(col_dpto) %>% 
  units::set_units("km^2") # Convert to square kilometers

road_lengths_per_dept <- roads_by_department %>%
  group_by(dpto_ccdgo) %>%  
  summarize(total_road_length_km = sum(length_km, na.rm = TRUE))

# Join departments with their respective road density
col_dpto <- col_dpto %>%
  left_join(st_drop_geometry(road_lengths_per_dept), by = "dpto_ccdgo") %>% 
  mutate(road_density = total_road_length_km / area_km2)


#===================================#
# 4. Population density dpto ------
#===================================#
if(!file.exists("datos/spatial/pop_dpto_intersect.rds")){
  # Prepare parallelization
  plan(multisession, workers = parallel::detectCores()-1)  
  
  intersections <- st_intersects(x = pop_hex, y = col_dpto)
  
  # Initialize progress bar
  pb <- progress::progress_bar$new(format = "[:bar] :current/:total (:percent)", total = dim(roads_colombia)[1])
  
  # Define parallelized intersection function
  intersectFeatures <- future_map_dfr(1:dim(pop_hex)[1], function(ix){
    pb$tick()
    st_intersection(x = pop_hex[ix,], y = col_dpto[intersections[[ix]],])
  })
  plan(sequential)  # Reset the plan to the default single-core mode
  saveRDS(intersectFeatures, 
          file ="datos/spatial/pop_dpto_intersect.rds" ) 
  pop_by_department <- intersectFeatures
  
} else{
  
  pop_by_department <- readRDS("datos/spatial/pop_dpto_intersect.rds")
}



# 5. Export -----
out <- col_dpto %>% 
  st_drop_geometry() %>% 
  select(dpto_ccdgo, area_km2, total_road_length_km, road_density, population_density)
arrow::write_parquet(out, "datos/spatial/colombia-departments_road_pop.parquet")

# End

File: codigo/descriptive/02_calculate_road_density_departments.py
# 1. Environment Setup
import geopandas as gpd
import pandas as pd
import numpy as np
from shapely.geometry import shape
import pyproj
import pyarrow as pa
import pyarrow.parquet as pq
from tqdm import tqdm
import multiprocessing

# 2. Data Import
# Load the spatial datasets
pop_hex = gpd.read_file("datos/population/colombia/kontur_population_CO_20231101.gpkg")
col_dpto = gpd.read_file("datos/spatial/MGN2023_DPTO_POLITICO/MGN_ADM_DPTO_POLITICO.shp")
roads_colombia = gpd.read_file("datos/spatial/colombia_roads.gpkg")

# 3. Coordinate Transformation
# Ensure all datasets use the same CRS (coordinate reference system)
if pop_hex.crs != col_dpto.crs:
    pop_hex = pop_hex.to_crs(col_dpto.crs)

if roads_colombia.crs != col_dpto.crs:
    roads_colombia = roads_colombia.to_crs(col_dpto.crs)

# 4. Spatial Intersections
# Helper function for spatial intersection using parallel processing
def parallel_intersection(df1, df2):
    # Split the data for parallel processing
    num_cores = multiprocessing.cpu_count()
    df_split = np.array_split(df1, num_cores)
    
    with multiprocessing.Pool(num_cores) as pool:
        # Use tqdm to track progress
        result = list(tqdm(pool.imap(lambda x: gpd.overlay(x, df2, how='intersection'), df_split), total=len(df_split)))
    
    # Concatenate the results back together
    return pd.concat(result, ignore_index=True)

# Roads with Departments
try:
    # If the file already exists, load it
    roads_dpto_intersect = gpd.read_parquet("datos/spatial/roads_dpto_intersect.rds")
except FileNotFoundError:
    # Perform the spatial intersection
    roads_dpto_intersect = parallel_intersection(roads_colombia, col_dpto)
    # Save to a file
    roads_dpto_intersect.to_parquet("datos/spatial/roads_dpto_intersect.rds", index=False)

# Population with Departments
try:
    # If the file already exists, load it
    pop_dpto_intersect = gpd.read_parquet("datos/spatial/pop_dpto_intersect.rds")
except FileNotFoundError:
    # Perform the spatial intersection
    pop_dpto_intersect = parallel_intersection(pop_hex, col_dpto)
    # Save to a file
    pop_dpto_intersect.to_parquet("datos/spatial/pop_dpto_intersect.rds", index=False)

# 5. Road Density Calculation
# Road Length Calculation
roads_dpto_intersect['length_km'] = roads_dpto_intersect.geometry.length / 1000  # Convert to kilometers

# Department Area Calculation
col_dpto['area_km2'] = col_dpto.geometry.area / 1e6  # Convert to square kilometers

# Aggregate Road Lengths
road_length_by_dpto = roads_dpto_intersect.groupby('dpto_ccdgo')['length_km'].sum().reset_index()
road_length_by_dpto.columns = ['dpto_ccdgo', 'total_road_length_km']

# Compute Road Density
col_dpto = col_dpto.merge(road_length_by_dpto, on='dpto_ccdgo', how='left')
col_dpto['road_density'] = col_dpto['total_road_length_km'] / col_dpto['area_km2']

# 6. Population Density Calculation
# Aggregate Population by Department
pop_by_dpto = pop_dpto_intersect.groupby('dpto_ccdgo')['population'].sum().reset_index()
pop_by_dpto.columns = ['dpto_ccdgo', 'total_population']

# Compute Population Density
col_dpto = col_dpto.merge(pop_by_dpto, on='dpto_ccdgo', how='left')
col_dpto['pop_density'] = col_dpto['total_population'] / col_dpto['area_km2']

# 7. Data Export
# Select and reorder the columns for final output
final_df = col_dpto[['dpto_ccdgo', 'area_km2', 'total_road_length_km', 'road_density', 'pop_density']].copy()

# Export the final DataFrame to a Parquet file
final_df.to_parquet('colombia-departments_road_pop.parquet', index=False)

